File: scripts/test_aarch64_tty_serial_atomic.py
#!/usr/bin/env python3
"""Exercise production serial output with host lock/PL011 adapters.

The production byte loops and formatted-output entry point are extracted, not
reimplemented. Only the AArch64 IRQ-mask/spinlock and MMIO byte writer are host
adapters. The real TTY crate supplies the previous output translation oracle.
This is a host atomicity regression, not an AArch64 hardware/runtime result.
"""

from pathlib import Path
import re
import subprocess
import tempfile


ROOT = Path(__file__).resolve().parents[1]
SERIAL = (ROOT / "kernel/src/serial.rs").read_text()
TTY = (ROOT / "kernel/src/aarch64_tty.rs").read_text()
ARCH = (ROOT / "kernel/src/arch/aarch64.rs").read_text()


def item(source: str, declaration: str) -> str:
    start = source.index(declaration)
    opening = source.index("{", start)
    depth = 1
    end = opening + 1
    while depth:
        depth += (source[end] == "{") - (source[end] == "}")
        end += 1
    return source[start:end]


def compact(source: str) -> str:
    return re.sub(r"\s+", "", re.sub(r"//[^\n]*", "", source))


tty_write = item(TTY, "pub fn write(fd:")
tty_compact = compact(tty_write)
serial_call = "crate::serial::write_tty_bytes(bytes,state.line.termios().output_crlf);"
assert tty_compact.count(serial_call) == 1
assert "letmutsink=TerminalDisplaySink;" in tty_compact
assert "state.line.write_output(bytes,&mutsink);" in tty_compact
assert tty_compact.index("returnErr(Errno::BadFileDescriptor);") < tty_compact.index(serial_call)
assert "!matches!(fd,1|2)||!state.fd_open(pid,fd)" in tty_compact
assert tty_compact.index(serial_call) < tty_compact.index("state.line.write_output")
assert "TerminalSink;" not in tty_write
assert "write_bytes(" not in tty_write
display = item(TTY, "impl ByteSink for TerminalDisplaySink")
assert "crate::graphics::terminal_write(bytes);" in display
assert "serial" not in display
echo = item(TTY, "impl ByteSink for TerminalSink")
assert "crate::serial::write_bytes(bytes);" in echo
assert "crate::graphics::terminal_write(bytes);" in echo

# The normal musl M_write=17 path must continue enforcing readable user memory
# and fd/TTY semantics. The record emitter does not bypass that route.
native_write_start = ARCH.index("        SYS_FILE_WRITE => {")
native_write = item(ARCH[native_write_start:], "        SYS_FILE_WRITE =>")
assert "user_range_readable(address, length)" in native_write
assert "crate::aarch64_tty::write(frame.registers[0], input)" in native_write

declarations = (
    "pub fn print(",
    "pub fn write_bytes(",
    "pub fn write_tty_bytes(",
    "fn write_output_bytes(",
    "impl Write for Serial",
)
excerpts = [item(SERIAL, declaration) for declaration in declarations]
output = item(SERIAL, "fn write_output_bytes(")
assert output.count("SerialGuard::acquire()") == 1
assert output.index("SerialGuard::acquire()") < output.index("for &byte in bytes")
assert "graphics" not in output
assert "write_output_bytes(bytes, false);" in excerpts[1]
assert "write_output_bytes(bytes, output_crlf);" in excerpts[2]

# Host tests substitute the lock, not the IRQ instructions. Keep the real
# acquire/release ordering protected structurally without executing host DAIF.
acquire = item(SERIAL, "impl SerialGuard")
release = item(SERIAL, "impl Drop for SerialGuard")
assert acquire.index('"msr daifset, #0xf"') < acquire.index("compare_exchange_weak")
assert "Ordering::Acquire" in acquire
assert release.index("SERIAL_LOCK.store(false, Ordering::Release)") < release.index('"msr daif, {saved}"')

production = "\n\n".join(excerpts).replace('#[cfg(target_arch = "aarch64")]\n', "")
assert "asm!(" not in production

with tempfile.TemporaryDirectory(prefix="makos-tty-serial-atomic-") as temporary:
    directory = Path(temporary)
    library = directory / "libmakos_tty.rlib"
    subprocess.run(
        ["rustc", "--edition=2024", "--crate-name", "makos_tty", "--crate-type", "rlib",
         str(ROOT / "crates/tty/src/lib.rs"), "-o", str(library)],
        check=True,
    )
    source = directory / "serial_output.rs"
    source.write_text(production)
    fixture = directory / "test.rs"
    fixture.write_text((ROOT / "scripts/test_aarch64_tty_serial_atomic.rs").read_text())

    def compile_fixture(name: str) -> Path:
        binary = directory / name
        subprocess.run(
            ["rustc", "--edition=2024", "--test", str(fixture),
             "--extern", f"makos_tty={library}", "-o", str(binary)],
            check=True,
        )
        return binary

    subprocess.run(
        [str(compile_fixture("tty-serial-tests")), "--test-threads=1"],
        check=True, timeout=30,
    )

    # Reintroduce the old production line-discipline-to-serial chunk boundary
    # only in the generated fixture. Its byte stream is unchanged but the
    # same complete-record critical-section assertion must reject it.
    old_tty_write = """
pub fn write_tty_bytes(bytes: &[u8], output_crlf: bool) {
    struct OldTerminalSink;
    impl makos_tty::ByteSink for OldTerminalSink {
        fn write(&mut self, bytes: &[u8]) { write_bytes(bytes); }
    }
    let mut termios = makos_tty::Termios::sane();
    termios.output_crlf = output_crlf;
    let line = makos_tty::LineDiscipline::<16, 8, 4>::new(termios);
    line.write_output(bytes, &mut OldTerminalSink);
}
"""
    source.write_text(production.replace(excerpts[2], old_tty_write))
    negative = subprocess.run(
        [str(compile_fixture("old-tty-chunks")), "--exact", "tty_record_has_one_serial_critical_section"],
        capture_output=True, text=True, timeout=15,
    )
    if negative.returncode == 0 or "TTY record split across serial critical sections" not in negative.stdout + negative.stderr:
        raise SystemExit(f"Old TTY boundary negative control did not fail correctly:\n{negative.stdout}{negative.stderr}")

    # Losing the outer guard must fail on the first attempted hardware byte,
    # even though a single-thread byte-equivalence check alone would pass.
    unguarded = output.replace('    #[cfg(target_arch = "aarch64")]\n', "")
    unguarded = unguarded.replace("    let _guard = SerialGuard::acquire();\n", "")
    adapted_output = output.replace('#[cfg(target_arch = "aarch64")]\n', "")
    source.write_text(production.replace(adapted_output, unguarded))
    negative = subprocess.run(
        [str(compile_fixture("unguarded-tty-output")), "--exact", "raw_serial_write_preserves_bytes"],
        capture_output=True, text=True, timeout=15,
    )
    if negative.returncode == 0 or "serial byte emitted without the outer guard" not in negative.stdout + negative.stderr:
        raise SystemExit(f"Missing serial guard negative control did not fail correctly:\n{negative.stdout}{negative.stderr}")

print("MAKOS_AARCH64_TTY_SERIAL_ATOMIC_HOST_OK output=production-code "
      "tty_translation=production-crate byte_equivalence=preserved "
      "record_lock=whole-write contender=kernel-formatted-log "
      "negative_controls=old-chunks,missing-guard adapters=host-mutex,byte-capture")

File: scripts/test_aarch64_tty_serial_atomic.rs
//! Host lock/PL011 adapters for the exact production serial output functions.
#![allow(dead_code)]

use core::fmt::{self, Write};
use makos_tty::{ByteSink, LineDiscipline, Termios};
use std::cell::Cell;
use std::sync::{Condvar, Mutex, MutexGuard};

include!("serial_output.rs");

static SERIAL_LOCK: Mutex<()> = Mutex::new(());
static CAPTURE: Mutex<Capture> = Mutex::new(Capture {
    bytes: Vec::new(),
    events: Vec::new(),
    next_guard: 0,
});
static GATE: Mutex<Gate> = Mutex::new(Gate {
    pause_at_first_byte: false,
    first_byte_seen: false,
    acquire_attempts: 0,
});
static GATE_CHANGED: Condvar = Condvar::new();

thread_local! {
    static ACTIVE_GUARD: Cell<usize> = const { Cell::new(0) };
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum Event {
    Acquire(usize),
    Byte(usize, u8),
    Release(usize),
}

struct Capture {
    bytes: Vec<u8>,
    events: Vec<Event>,
    next_guard: usize,
}

struct Gate {
    pause_at_first_byte: bool,
    first_byte_seen: bool,
    acquire_attempts: usize,
}

struct SerialGuard {
    id: usize,
    _lock: MutexGuard<'static, ()>,
}

impl SerialGuard {
    fn acquire() -> Self {
        assert_eq!(ACTIVE_GUARD.get(), 0, "recursive serial lock acquisition");
        {
            let mut gate = GATE.lock().unwrap();
            gate.acquire_attempts += 1;
            GATE_CHANGED.notify_all();
        }
        let lock = SERIAL_LOCK.lock().unwrap();
        let id = {
            let mut capture = CAPTURE.lock().unwrap();
            capture.next_guard += 1;
            let id = capture.next_guard;
            capture.events.push(Event::Acquire(id));
            id
        };
        ACTIVE_GUARD.set(id);
        Self { id, _lock: lock }
    }
}

impl Drop for SerialGuard {
    fn drop(&mut self) {
        assert_eq!(ACTIVE_GUARD.get(), self.id);
        CAPTURE.lock().unwrap().events.push(Event::Release(self.id));
        ACTIVE_GUARD.set(0);
    }
}

struct Serial;

impl Serial {
    fn write_byte(&mut self, byte: u8) {
        let id = ACTIVE_GUARD.get();
        assert_ne!(id, 0, "serial byte emitted without the outer guard");
        {
            let mut capture = CAPTURE.lock().unwrap();
            capture.bytes.push(byte);
            capture.events.push(Event::Byte(id, byte));
        }
        let mut gate = GATE.lock().unwrap();
        if gate.pause_at_first_byte && !gate.first_byte_seen {
            gate.first_byte_seen = true;
            GATE_CHANGED.notify_all();
            // Force a second real host thread to attempt the same serial lock
            // while the first record owns it. This is deterministic contention,
            // not a sleep or a claim to execute AArch64 IRQ masking on the host.
            while gate.acquire_attempts < 2 {
                gate = GATE_CHANGED.wait(gate).unwrap();
            }
        }
    }
}

fn reset() {
    assert_eq!(ACTIVE_GUARD.get(), 0);
    let _serial = SERIAL_LOCK.lock().unwrap();
    *CAPTURE.lock().unwrap() = Capture {
        bytes: Vec::new(),
        events: Vec::new(),
        next_guard: 0,
    };
    *GATE.lock().unwrap() = Gate {
        pause_at_first_byte: false,
        first_byte_seen: false,
        acquire_attempts: 0,
    };
}

/// Reference the actual line discipline, followed by the pre-existing serial
/// LF translation. Deliberately preserve the historical ONLCR CRCRLF bytes.
fn old_tty_bytes(bytes: &[u8], output_crlf: bool) -> Vec<u8> {
    struct OldSerialSink(Vec<u8>);
    impl ByteSink for OldSerialSink {
        fn write(&mut self, bytes: &[u8]) {
            for &byte in bytes {
                if byte == b'\n' {
                    self.0.push(b'\r');
                }
                self.0.push(byte);
            }
        }
    }
    let mut termios = Termios::sane();
    termios.output_crlf = output_crlf;
    let line = LineDiscipline::<16, 8, 4>::new(termios);
    let mut sink = OldSerialSink(Vec::new());
    line.write_output(bytes, &mut sink);
    sink.0
}

fn assert_guarded_records(expected_records: usize) {
    let capture = CAPTURE.lock().unwrap();
    assert_eq!(
        capture.next_guard, expected_records,
        "TTY record split across serial critical sections"
    );
    let mut active = None;
    let mut releases = 0;
    for event in &capture.events {
        match *event {
            Event::Acquire(id) => {
                assert_eq!(active, None, "overlapping serial ownership");
                active = Some(id);
            }
            Event::Byte(id, _) => assert_eq!(active, Some(id)),
            Event::Release(id) => {
                assert_eq!(active, Some(id));
                active = None;
                releases += 1;
            }
        }
    }
    assert_eq!(active, None);
    assert_eq!(releases, expected_records);
}

const RECORD: &[u8] = b"MAKOS_MUSL_EL0_ENTRY_OK loader=musl threads=3 singleton=0x2,0x4,0x8 tids=5,6,7 pthread_create=0x280afcd4 rx=0x80000000 resume=0x80001000 calls=96 statuses=42,42,42 block=sleep-until\n";

#[test]
fn tty_record_has_one_serial_critical_section() {
    reset();
    assert!(RECORD.len() <= 512);
    write_tty_bytes(RECORD, true);
    assert_eq!(CAPTURE.lock().unwrap().bytes, old_tty_bytes(RECORD, true));
    assert_guarded_records(1);
}

#[test]
fn tty_output_preserves_existing_translation_for_all_bytes() {
    let all_bytes: Vec<u8> = (0..=255).collect();
    let mut bounded_record = vec![b'x'; 511];
    bounded_record.push(b'\n');
    let multiline = b"first\nsecond\r\n\nlast\r".repeat(300);
    for output_crlf in [false, true] {
        for bytes in [
            &b""[..],
            &b"\n"[..],
            &b"\r\n"[..],
            &b"ordinary text without newline"[..],
            &all_bytes,
            &bounded_record,
            &multiline,
        ] {
            reset();
            write_tty_bytes(bytes, output_crlf);
            assert_eq!(
                CAPTURE.lock().unwrap().bytes,
                old_tty_bytes(bytes, output_crlf)
            );
            assert_guarded_records(1);
        }
    }
}

#[test]
fn raw_serial_write_preserves_bytes() {
    reset();
    write_bytes(b"one\r\ntwo\n\0\xff");
    assert_eq!(CAPTURE.lock().unwrap().bytes, b"one\r\r\ntwo\r\n\0\xff");
    assert_guarded_records(1);
}

#[test]
fn kernel_formatted_record_uses_same_guard() {
    reset();
    print(format_args!("kernel cpu={} tid={} status={}\n", 1, 5, 42));
    assert_eq!(
        CAPTURE.lock().unwrap().bytes,
        b"kernel cpu=1 tid=5 status=42\r\n"
    );
    assert_guarded_records(1);
}

#[test]
fn fatal_detail_precedes_stop_marker_under_one_guard() {
    reset();
    fatal("synthetic host-test reason");
    let bytes = CAPTURE.lock().unwrap().bytes.clone();
    let marker = b"MAKOS_FATAL:";
    let detail = b"MAKOS_FAILURE_DETAIL: synthetic host-test reason\r\n";
    // Simulate every possible two-chunk boundary, including immediately
    // after the colon. Detection is still an immediate prefix predicate.
    for split in 0..=bytes.len() {
        let mut captured = bytes[..split].to_vec();
        if !captured.windows(marker.len()).any(|part| part == marker) {
            captured.extend_from_slice(&bytes[split..]);
        }
        let stop = captured.windows(marker.len()).position(|part| part == marker).unwrap();
        assert!(
            captured[..stop].ends_with(detail),
            "fatal reason missing before immediate stop marker at split {split}"
        );
    }
    assert_eq!(
        bytes,
        b"MAKOS_FAILURE_DETAIL: synthetic host-test reason\r\nMAKOS_FATAL: synthetic host-test reason\r\n"
    );
    assert_guarded_records(1);
}

#[test]
fn concurrent_fatal_records_keep_their_own_detail() {
    reset();
    GATE.lock().unwrap().pause_at_first_byte = true;
    let writer = std::thread::spawn(|| fatal("first synthetic reason"));
    {
        let mut gate = GATE.lock().unwrap();
        while !gate.first_byte_seen {
            gate = GATE_CHANGED.wait(gate).unwrap();
        }
    }
    let contender = std::thread::spawn(|| fatal("second synthetic reason"));
    writer.join().unwrap();
    contender.join().unwrap();
    assert_eq!(
        CAPTURE.lock().unwrap().bytes,
        b"MAKOS_FAILURE_DETAIL: first synthetic reason\r\nMAKOS_FATAL: first synthetic reason\r\nMAKOS_FAILURE_DETAIL: second synthetic reason\r\nMAKOS_FATAL: second synthetic reason\r\n"
    );
    assert_guarded_records(2);
    assert_eq!(GATE.lock().unwrap().acquire_attempts, 2);
}

#[test]
fn concurrent_kernel_trace_cannot_split_tty_record() {
    for output_crlf in [false, true] {
        reset();
        GATE.lock().unwrap().pause_at_first_byte = true;
        let writer = std::thread::spawn(move || write_tty_bytes(RECORD, output_crlf));
        {
            let mut gate = GATE.lock().unwrap();
            while !gate.first_byte_seen {
                gate = GATE_CHANGED.wait(gate).unwrap();
            }
        }
        let contender = std::thread::spawn(|| {
            print(format_args!(
                "MAKOS_AARCH64_THREAD_EXIT_OK tid={} status={}\n",
                7, 0
            ));
        });
        writer.join().unwrap();
        contender.join().unwrap();
        let mut expected = old_tty_bytes(RECORD, output_crlf);
        expected.extend_from_slice(b"MAKOS_AARCH64_THREAD_EXIT_OK tid=7 status=0\r\n");
        assert_eq!(
            CAPTURE.lock().unwrap().bytes,
            expected,
            "kernel trace spliced into TTY record"
        );
        assert_guarded_records(2);
        assert_eq!(GATE.lock().unwrap().acquire_attempts, 2);
    }
}

#[test]
fn successive_terminal_records_have_distinct_atomic_boundaries() {
    reset();
    for _ in 0..3 {
        write_tty_bytes(RECORD, true);
    }
    assert_eq!(
        CAPTURE.lock().unwrap().bytes,
        old_tty_bytes(RECORD, true).repeat(3)
    );
    assert_guarded_records(3);
}
